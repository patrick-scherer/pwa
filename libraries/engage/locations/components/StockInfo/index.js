// @flow
export { StockInfo } from './StockInfo';
