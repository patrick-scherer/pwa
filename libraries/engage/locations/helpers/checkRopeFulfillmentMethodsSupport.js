import { ROPIS, BOPIS } from '../constants';

/**
 * Checks if an fulfillment methods array contains supported fulfillment methods.
 * @param {Array} fulfillmentMethods An array of product fulfillment methods.
 * @return {boolean}
 */
const checkRopeFulfillmentMethodsSupport = (fulfillmentMethods = []) => {
  if (!Array.isArray(fulfillmentMethods)) {
    return false;
  }

  return fulfillmentMethods.includes(ROPIS) || fulfillmentMethods.includes(BOPIS);
};

export default checkRopeFulfillmentMethodsSupport;
