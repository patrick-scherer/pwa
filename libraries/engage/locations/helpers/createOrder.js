// @flow
import {
  getCurrency, getCartItems, getSubTotal, getGrandTotal,
} from '@shopgate/pwa-common-commerce/cart/selectors';
import { getProductDataById } from '@shopgate/pwa-common-commerce/product/selectors/product';
import { makeGetUserLocation } from '../selectors';

/**
 * Creates the address sequence.
 * @param {Object} formValues The reserve form values.
 * @returns {Array}
 */
function createAddressSequence(formValues) {
  return [
    {
      type: 'billing',
      firstName: formValues.firstName,
      lastName: formValues.lastName,
      phone: formValues.cellPhone,
      emailAddress: formValues.email,
    },
    {
      type: 'pickup',
      firstName: formValues.firstName2,
      lastName: formValues.lastName2,
      phone: formValues.cellPhone2,
      emailAddress: formValues.email2,
    },
  ];
}

/**
 * Retrieves the featured image for a product. If the product doesn't have a featured image
 * and is a child product, the featured image of the base product is retrieved.
 * @param {Object} product The current product.
 * @param {FUnction} getState The redux getState function.
 * @returns {string}
 */
function getProductImage(product, getState) {
  let image = product.featuredImageUrl;

  if (!image && product.baseProductId) {
    const baseProduct = getProductDataById(getState(), { productId: product.id });

    if (baseProduct) {
      image = baseProduct.featuredImageUrl;
    }
  }

  return image;
}

/**
 * Creates a single order line item (from quick reserve).
 * @param {Object} product The current product.
 * @param {Function} getState The redux getState function.
 * @returns {Array}
 */
function createSingleProductItems(product, getState) {
  const location = makeGetUserLocation()(getState());

  return [
    {
      code: product.id,
      quantity: 1,
      fulfillmentMethod: 'ROPIS', // TODO: Change based on the selected method.
      fulfillmentLocationCode: location.code,
      price: product.price.unitPrice,
      shipToAddressSequenceIndex: 0,
      currencyCode: product.price.currency,
      product: {
        code: product.id,
        name: product.name,
        image: getProductImage(product, getState),
        price: product.price.unitPrice,
        currencyCode: product.price.currency,
      },
    },
  ];
}

/**
 * Creates the order line items.
 * @param {Function} getState The redux getState function.
 * @returns {Array}
 */
function createCartLineItems(getState) {
  const items = getCartItems(getState());
  const currencyCode = getCurrency(getState());

  return items.map(item => ({
    code: item.id,
    quantity: item.quantity,
    fulfillmentMethod: item.fulfillment.method,
    fulfillmentLocationCode: item.fulfillment.location.code,
    price: item.product.price.unit,
    shipToAddressSequenceIndex: 0,
    currencyCode,
    product: {
      code: item.product.id,
      name: item.product.name,
      image: item.product.featuredImageUrl,
      price: item.product.price.unit,
      currencyCode,
    },
  }));
}

/**
 * Creates a fulfillment order out of the reserve form values and the current product or the cart.
 * @param {Object} formValues The reserve form values.
 * @param {Object} product The current product.
 * @param {Function} getState The redux getState function.
 * @returns {Object}
 */
function createOrder(formValues: { [string]: string }, product: any, getState: () => any) {
  // If no individual product was submitted, we handle the cart.
  if (product === null) {
    return {
      // localeCode: process.env.LOCALE_FILE.replace('-', '_'),
      localeCode: 'de_DE',
      currencyCode: getCurrency(getState()),
      addressSequences: createAddressSequence(formValues),
      primaryBillToAddressSequenceIndex: 0,
      lineItems: createCartLineItems(getState),
      subTotal: getSubTotal(getState()),
      total: getGrandTotal(getState()),
    };
  }

  return {
    // localeCode: process.env.LOCALE_FILE.replace('-', '_'),
    localeCode: 'de_DE',
    currencyCode: product.price.currency,
    addressSequences: createAddressSequence(formValues),
    primaryBillToAddressSequenceIndex: 0,
    lineItems: createSingleProductItems(product, getState),
    subTotal: product.price.unitPrice,
    total: product.price.unitPrice,
  };
}

export default createOrder;
