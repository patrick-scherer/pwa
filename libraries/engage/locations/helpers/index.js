// @flow
export { default as createOrder } from './createOrder';
export { default as getAvailabilitySettings } from './getAvailabilitySettings';
export { default as checkRopeFulfillmentMethodsSupport } from './checkRopeFulfillmentMethodsSupport';
export { default as getDefaultRopeFulfillmentMethod } from './getDefaultRopeFulfillmentMethod';
export { isProductAvailable } from './productInventory';
