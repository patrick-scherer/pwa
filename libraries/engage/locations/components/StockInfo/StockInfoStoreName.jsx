import React from 'react';
import PropTypes from 'prop-types';
import { capitalize } from '../../../core';
import { I18n } from '../../../components';

/**
 * Renders the given store name if one is set and capitalizes the resulting text if asked for.
 * @param {string} props.name The name to display.
 * @param {Object} props.displayCapitalized If set, the translated text will be capitalized.
 * @return {JSX}
 */
function StoreName({ name, displayCapitalized }) {
  if (!name) {
    return null;
  }

  return (
    <I18n.Text
      string="locations.stock_info.pick_up_at"
      params={{ storeName: name }}
      transform={displayCapitalized ? capitalize : null}
    />
  );
}

StoreName.propTypes = {
  displayCapitalized: PropTypes.bool,
  name: PropTypes.string,
};

StoreName.defaultProps = {
  displayCapitalized: false,
  name: '',
};

export default StoreName;
