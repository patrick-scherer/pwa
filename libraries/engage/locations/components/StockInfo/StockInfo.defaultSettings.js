import { themeConfig } from '@shopgate/pwa-common/helpers/config';
import {
  AVAILABILITY_TYPE_AVAILABLE,
  AVAILABILITY_TYPE_LIMITED_AVAILABILITY,
  AVAILABILITY_TYPE_NOT_AVAILABLE,
  AVAILABILITY_TYPE_CUSTOM_AVAILABILITY,
} from '../../../product/constants';

const { colors } = themeConfig;

export default {
  [AVAILABILITY_TYPE_AVAILABLE]: {
    visibleInventoryFrom: 11,
    visibleInventoryTo: null,
    includeInventoryBlind: false,
    includeNoRecordStores: false,
    availabilityText: `locations.stock_info.${AVAILABILITY_TYPE_AVAILABLE}`,
    availabilityTextColor: colors.success,
  },
  [AVAILABILITY_TYPE_LIMITED_AVAILABILITY]: {
    visibleInventoryFrom: 1,
    visibleInventoryTo: 10,
    includeInventoryBlind: false,
    includeNoRecordStores: false,
    availabilityText: `locations.stock_info.${AVAILABILITY_TYPE_LIMITED_AVAILABILITY}`,
    availabilityTextColor: colors.warning,
  },
  [AVAILABILITY_TYPE_NOT_AVAILABLE]: {
    visibleInventoryFrom: 0,
    visibleInventoryTo: 0,
    includeInventoryBlind: false,
    includeNoRecordStores: false,
    availabilityText: `locations.stock_info.${AVAILABILITY_TYPE_NOT_AVAILABLE}`,
    availabilityTextColor: colors.error,
    allowOrderingToStore: true,
  },
  [AVAILABILITY_TYPE_CUSTOM_AVAILABILITY]: {
    visibleInventoryFrom: 0,
    visibleInventoryTo: 0,
    includeInventoryBlind: false,
    includeNoRecordStores: true,
    availabilityText: `locations.stock_info.${AVAILABILITY_TYPE_CUSTOM_AVAILABILITY}`,
    availabilityTextColor: colors.error,
    allowOrderingToStore: true,
  },
  maxNumberOfVisibleInventory: 100,
  aboveMaxExtension: '++',
};
