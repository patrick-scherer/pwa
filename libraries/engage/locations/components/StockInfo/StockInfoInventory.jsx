// @flow
import * as React from 'react';
import { I18n } from '../../../components';
import { type Location } from '../../locations.types';

type Props = {
  availabilityText: string,
  location: Location,
  maxNumberVisible: number,
  aboveMaxExtension: string,
}

/**
 * Renders the inventory given by the location into a given translation string.
 * @param {Object} props The component props.
 * @param {string} props.availabilityText The translation string to use.
 * @param {Object} props.location The component props.
 * @param {number|undefined} props.maxNumberVisible The component props.
 * @param {string} props.aboveMaxExtension The component props.
 * @return {JSX}
 */
export function StockInfoInventory(props: Props) {
  const {
    availabilityText, location, maxNumberVisible, aboveMaxExtension,
  } = props;

  const visibleInventory = React.useMemo(() => {
    if (!location || !location.productInventory) {
      return null;
    }

    if (
      location.productInventory.visible !== null
      && location.productInventory.visible > maxNumberVisible
    ) {
      return `${maxNumberVisible}${aboveMaxExtension}`;
    }
    return location.productInventory.visible;
  }, [aboveMaxExtension, location, maxNumberVisible]);

  if (!location || !availabilityText || !location.productInventory) {
    return null;
  }

  return (
    <I18n.Text string={availabilityText} params={{ visibleInventory }} />
  );
}
