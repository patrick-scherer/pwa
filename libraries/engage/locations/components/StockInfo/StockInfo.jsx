// @flow
import * as React from 'react';
import { css } from 'glamor';
import classNames from 'classnames';
import defaultsDeep from 'lodash/defaultsDeep';
import { getThemeSettings } from '../../../core';
import { SurroundPortals } from '../../../components';
import { PRODUCT_LOCATION_STOCK_INFO } from '../../constants';
import { getAvailabilitySettings } from '../../helpers';
import defaultSettings from './StockInfo.defaultSettings';
import { StockInfoInventory } from './StockInfoInventory';
import StoreName from './StockInfoStoreName';
import { type Location } from '../../locations.types';

type Props = {
  location: Location,
  className?: string | {} | null,
  showStoreName?: boolean,
}

/**
 * Renders visible stock information based on the given location.
 * @param {Object} props The component props.
 * @return {JSX}
 */
export function StockInfo({ location, className, showStoreName }: Props) {
  const { locationStockInfo } = getThemeSettings('product') || {};
  const settings = defaultsDeep(locationStockInfo, defaultSettings);

  const { availabilityText = '', availabilityTextColor = 'inherit' } =
    getAvailabilitySettings(settings, location);

  const defaultClassName = css({
    color: availabilityTextColor,
    fontSize: '0.875rem',
    margin: 0,
  }).toString();

  const displayCapitalized = React.useMemo(() => (
    location.productInventory && location.productInventory.visible === null
  ) || !availabilityText, [availabilityText, location.productInventory]);

  const portalProps = React.useMemo(() => ({
    location,
    className,
    availabilityText,
    availabilityTextColor,
  }), [availabilityText, availabilityTextColor, className, location]);

  return (
    <SurroundPortals portalName={PRODUCT_LOCATION_STOCK_INFO} portalProps={portalProps}>
      <span className={classNames(defaultClassName, css(className).toString())}>
        <StockInfoInventory
          availabilityText={availabilityText}
          location={location}
          maxNumberVisible={settings.maxNumberOfVisibleInventory}
          aboveMaxExtension={settings.aboveMaxExtension}
        />
        <StoreName
          name={showStoreName ? location.name : null}
          displayCapitalized={displayCapitalized}
        />
      </span>
    </SurroundPortals>
  );
}

StockInfo.defaultProps = {
  className: null,
  showStoreName: true,
};
