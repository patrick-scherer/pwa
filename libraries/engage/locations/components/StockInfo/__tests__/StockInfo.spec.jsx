import React from 'react';
import { mount } from 'enzyme';
import { getThemeSettings } from '../../../../core';
import { getAvailabilitySettings } from '../../../helpers';
import { StockInfo } from '../StockInfo';

jest.mock('../../../../core', () => ({
  ...jest.requireActual('../../../../core'),
  getThemeSettings: jest.fn(),
}));
jest.mock('../../../helpers', () => ({
  getAvailabilitySettings: jest.fn(),
}));

describe('<StockInfo />', () => {
  it('should render with the visible inventory and with the store name', () => {
    getThemeSettings.mockReturnValueOnce({});
    getAvailabilitySettings.mockReturnValueOnce({
      availabilityText: 'demo_availability_text',
      availabilityTextColor: 'green',
    });
    const wrapper = mount(<StockInfo
      location={{
        name: 'Test Store',
      }}
    />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render without store name', () => {
    getThemeSettings.mockReturnValueOnce({});
    getAvailabilitySettings.mockReturnValueOnce({
      availabilityText: 'demo_availability_text',
      availabilityTextColor: 'green',
    });
    const wrapper = mount(<StockInfo
      location={{
        name: 'Test Store',
      }}
      showStoreName={false}
    />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('[string="product.location_stock_info.pick_up_at"]').length).toBe(0);
  });

  it('should render without the visible inventory', () => {
    getThemeSettings.mockReturnValueOnce({});
    getAvailabilitySettings.mockReturnValueOnce({
      availabilityText: 'demo_availability_text',
      availabilityTextColor: 'green',
    });
    const wrapper = mount(<StockInfo
      location={{
        name: 'Test Store',
      }}
    />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('[string="demo_availability_text"]').length).toBe(0);
  });
});
