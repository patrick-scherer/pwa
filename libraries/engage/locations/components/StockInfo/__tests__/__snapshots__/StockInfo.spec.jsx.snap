// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StockInfo /> should render with the visible inventory and with the store name 1`] = `
<StockInfo
  className={null}
  location={
    Object {
      "name": "Test Store",
    }
  }
>
  <SurroundPortals
    portalName="product.location-stock-info"
    portalProps={
      Object {
        "availabilityText": "demo_availability_text",
        "availabilityTextColor": "green",
        "className": null,
        "location": Object {
          "name": "Test Store",
        },
      }
    }
  >
    <Portal
      name="product.location-stock-info.before"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": null,
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
    <Portal
      name="product.location-stock-info"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": null,
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    >
      <span
        className="css-1604zsc css-nil"
      >
        <StockInfoInventory
          aboveMaxExtension="++"
          availabilityText="demo_availability_text"
          location={
            Object {
              "name": "Test Store",
            }
          }
          maxNumberVisible={100}
        />
      </span>
    </Portal>
    <Portal
      name="product.location-stock-info.after"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": null,
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
  </SurroundPortals>
</StockInfo>
`;

exports[`<StockInfo /> should render without store name 1`] = `
<StockInfo
  className={null}
  location={
    Object {
      "name": "Test Store",
    }
  }
  showStoreName={false}
>
  <SurroundPortals
    portalName="product.location-stock-info"
    portalProps={
      Object {
        "availabilityText": "demo_availability_text",
        "availabilityTextColor": "green",
        "className": null,
        "location": Object {
          "name": "Test Store",
        },
      }
    }
  >
    <Portal
      name="product.location-stock-info.before"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": null,
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
    <Portal
      name="product.location-stock-info"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": null,
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    >
      <span
        className="css-1604zsc css-nil"
      >
        <StockInfoInventory
          aboveMaxExtension="++"
          availabilityText="demo_availability_text"
          location={
            Object {
              "name": "Test Store",
            }
          }
          maxNumberVisible={100}
        />
      </span>
    </Portal>
    <Portal
      name="product.location-stock-info.after"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": null,
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
  </SurroundPortals>
</StockInfo>
`;

exports[`<StockInfo /> should render without the visible inventory 1`] = `
<StockInfo
  className={null}
  location={
    Object {
      "name": "Test Store",
    }
  }
>
  <SurroundPortals
    portalName="product.location-stock-info"
    portalProps={
      Object {
        "availabilityText": "demo_availability_text",
        "availabilityTextColor": "green",
        "className": null,
        "location": Object {
          "name": "Test Store",
        },
      }
    }
  >
    <Portal
      name="product.location-stock-info.before"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": null,
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
    <Portal
      name="product.location-stock-info"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": null,
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    >
      <span
        className="css-1604zsc css-nil"
      >
        <StockInfoInventory
          aboveMaxExtension="++"
          availabilityText="demo_availability_text"
          location={
            Object {
              "name": "Test Store",
            }
          }
          maxNumberVisible={100}
        />
      </span>
    </Portal>
    <Portal
      name="product.location-stock-info.after"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": null,
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
  </SurroundPortals>
</StockInfo>
`;
